src/Control/Middleware/RateLimitMiddleware.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+namespace SilverStripe\Control\Middleware;
+
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Core\Cache\RateLimiter;
+use SilverStripe\ErrorPage\ErrorPage;
+use SilverStripe\ORM\FieldType\DBDatetime;
+use SilverStripe\Security\Security;
+
+class RateLimitMiddleware implements HTTPMiddleware {
+
+    /**
+     * @param HTTPRequest $request
+     * @param callable $delegate
+     * @return HTTPResponse
+     */
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        $limiter = new RateLimiter($this->getKeyFromRequest($request), 10, 1);
+        if ($limiter->canAccess()) {
+            $limiter->hit();
+            $response = $delegate($request);
+        } else {
+            $response = $this->getErrorHTTPResponse();
+        }
+        $this->addHeadersToResponse($response, $limiter);
+        return $response;
+    }
+
+    /**
+     * @param HTTPRequest $request
+     * @return string
+     */
+    protected function getKeyFromRequest($request)
+    {
+        $domain = parse_url($request->getURL(), PHP_URL_HOST);
+        if ($currentUser = Security::getCurrentUser()) {
+            return md5($domain . '-' . $currentUser->ID);
+        }
+        return md5($domain . '-' . $request->getIP());
+    }
+
+    /**
+     * @return HTTPResponse
+     */
+    protected function getErrorHTTPResponse()
+    {
+        $response = null;
+        if (class_exists(ErrorPage::class)) {
```

Comment:
I'm not sure if it's worth adding this. Perhaps the error page module should have it's own middleware that reformats error responses rather than coupling like this...

It also doesn't work if there's no error page in the CMS for the 429 response anyway

Reply:
I'm removing this